README
# Struggle for Balance time convertor
Hello this is my little project for my universe which allows to convert time and dates from Earth Standard Time to Universal Standard Time which is the common time unit in my fictional universe.

## Background
The times share one second as the common unit, above that the UST follows this formula:

1 second

Standard minute = 100 seconds

Standard hour = 100 Standard minutes

Standard day = 10 Standard hours

Standard year = 500 Standard days

### UST
UST was established on new years eve of Earth Year 2400 to celebrate a new era in the human development and in the galactic community as well. Also the joining of Earth Confederacy gave enough votes to push the proposal through. It were the Grimnaar and Iyat'ktal who were pro and the Tro'ael and Axxonians who were against it, or to be more precise the Axxonians were indifferent. The Grimnaar and Iyat'ktal wanted order in times as previously it meant that every outpost had its own time and that made time synchronization very difficult. When humans joined in they joined the Grimnaar and Iyat'ktal as they already had similar experiences with their own outposts and colonies. The Tro'ael saw independent times as expressions of identity and the Axxoni were still on old Alliance time from thousands of years ago.

### EST
To make things complicated starting in 2400 there was a calendar reform that made February 29th not a transition year anomaly, but a standard day to be part of the calendar every year. This was to account for the changes in Earths orbit due to events previous events.

### IST
Imperial Standard Time is a time unit that was used by the Grimnaar and begins dating from the establishment of the united empire. Often times Grimnaar still refer to event in "before empire" and "during empire".

## Future plans
The plans are to improve the convertor with few more features described in the issues and hopefully later add more times like the mentioned IST.


changelog.md
# CHANGELOG
## v2.1 - [in development]
### Additions
* Increased test coverage
* Changelog
* How to use in Readme

### Fixes
* Incorrect starting time creation based on local time instead of on UTC

## v2.0 - released June 7 2015
### Changes
* The main part of the code was separated into an independent library

## v1.0.1 - released April 9 2015
### Additions
* Add link to the universe database

### Fixes
* Typos
* Readme look
* #1 displaying 100 instead of 00 in current SUT time

## v1.0 - released March 26 2015
* the initial version


js/tests.js
//create fictionalTime
var unit1 = 1000; //how many miliseconds does the unit have
var unit2 = unit1 * 100;
var unit3 = unit2 * 100;
var unit4 = unit3 * 10;
var unit5 = unit4 * 500;

var unitArray = [unit5, unit4, unit3, unit2, unit1]; //must be in order from the largest to smallest
var unitSeparator = ["SUT ", ".", " ", ":", ":", ""]; //formating, determines what symbols will be between different
var unitDeclaration = "before"; //options: before, after, both, false - additional option for unitSeparator to show time declaration like SUT xxxx

var timeBegins = moment("2400/01/01 00:00 GMT"); //the moment when time begins in relation to Earth time

QUnit.test("Correct starting time", function( assert ){
  assert.deepEqual(timeBegins.valueOf(), 13569465600000);
});

//create the object
var sut = new fictionalTime("Standard Universal Time", unitArray, timeBegins.valueOf(), unitSeparator, unitDeclaration);

//fictionalTime tests
QUnit.test("toTime", function( assert ){
  assert.deepEqual(sut.toTime(10000), "SUT 0.000 0:00:10");
  assert.deepEqual(sut.toTime(100000), "SUT 0.000 0:01:00");
  assert.deepEqual(sut.toTime(10000000), "SUT 0.000 1:00:00");
  assert.deepEqual(sut.toTime(100000000), "SUT 0.001 0:00:00");
  assert.deepEqual(sut.toTime(50000000000), "SUT 1.000 0:00:00");
  assert.deepEqual(sut.toTime(100000000000), "SUT 2.000 0:00:00");
  assert.deepEqual(sut.toTime(550000000000), "SUT 11.000 0:00:00");
  assert.deepEqual(sut.toTime(10550000000000), "SUT 211.000 0:00:00");
});

QUnit.test("toDate",  function( assert ){
  assert.deepEqual(sut.toDate(10000), "SUT -271.306 4:45:10", "Calculate date of 1970-01-01T00:00:10.000Z in SUT");
  assert.deepEqual(sut.toDate(30000000000000), "SUT 328.305 3:44:00", "Calculate date of 2920-08-30T05:20:00.000Z in SUT");
});

QUnit.test("toFictionalUnit", function( assert ){
  assert.deepEqual(sut.toFictionalUnit(10000, 4), 10, "SUT seconds");
  assert.deepEqual(sut.toFictionalUnit(100000, 3), 1, "SUT minutes");
  assert.deepEqual(sut.toFictionalUnit(10000000, 2), 1, "SUT hours");
  assert.deepEqual(sut.toFictionalUnit(100000000, 1), 1, "SUT years");
  assert.deepEqual(sut.toFictionalUnit(550000000000, 0), 11, "SUT years");
  assert.deepEqual(sut.toFictionalUnit(10550000000000, 0), 211, "SUT years");
});

QUnit.test("unitToMilliseconds", function( assert ){
  assert.deepEqual(sut.unitToMilliseconds(10, 4), 10000, "10 SUT seconds -> milliseconds");
  assert.deepEqual(sut.unitToMilliseconds(1, 3), 100000, "1 SUT minute -> milliseconds");
  assert.deepEqual(sut.unitToMilliseconds(10, 2), 100000000, "10 SUT hours -> milliseconds");
  assert.deepEqual(sut.unitToMilliseconds(1, 0), 50000000000, "SUT year -> milliseconds");
  assert.deepEqual(sut.unitToMilliseconds(211, 0), 10550000000000, "211 SUT years -> milliseconds");
});

/** @todo devise a test for sut.currentTime() */

//main page converters testing
QUnit.test("SUTyearsToEarth", function ( assert ){
  assert.equal(SUTyearsToEarth(15.8112), 25);
});

QUnit.test("earthYearsToSUT", function ( assert ){
  assert.equal(earthYearsToSUT(25), 15.8112);
});

QUnit.test("earthDateToSUT", function( assert ){
  assert.deepEqual(earthDateToSUT("2300-07-23"), "SUT -62.119 7:56:00", "Before SUT");
  assert.deepEqual(earthDateToSUT("2400-07-23"), "SUT 0.176 2:56:00", "SUT date");
});

QUnit.test("sutDateToEY", function( assert ){
  assert.deepEqual(sutDateToEY("000.000"), "2400-01-01", "First day of the new calendar");
  assert.deepEqual(sutDateToEY("12.000"), "2419-01-05", "SUT");
  assert.deepEqual(sutDateToEY("-12.490"), "2382-07-16", "Before SUT");
  assert.deepEqual(sutDateToEY("bleh"), "Please enter a valid SUT date in this format: YYY.DDD", "Invalid entry");
  assert.deepEqual(sutDateToEY("12"), "Please enter a valid SUT date in this format: YYY.DDD", "Only year");
  assert.deepEqual(sutDateToEY("12.650"), "SUT Year has only 500 days. Which means you should enter days between 000 to 499.", "More than 500 days in a year");
});


js/load.js
var time;
$(function(){
  /**
   * Load the submited time
   */
  $("#load").submit(function(event){
    event.preventDefault();

    //start animation on the button
    $("#cog").attr("class", "fa fa-cog fa-spin");

    //parse into JSON
    var code = $("#timeCode").val();

    var valid = true;

    var json, name, beginning, units, separators, declaration;
    try {
      json = JSON.parse(code);

      //separate all the parts into individual variables
      name = json.name;
      beginning = json.beginning;
      units = json.units;
      separators = json.separators;
      declaration = json.declaration;

    } catch (e) {
      console.log(e);
      //display error after #warning
      $("#warning").after('<div id="warning" data-alert class="alert-box alert radius">Incorrect JSON format.<a href="#" class="close">&times;</a></div>');
      valid = false;
    }

    if(valid)
    {
      /**
       * Validate the variables
       */
      //name - check for any potential code insertions

      //beginning - make sure that it is an int

      //units - check that it is a int array
      if( !Array.isArray(units) )
      {
        valid = false;
      }

      //separators - check that it is a proper string array
      //check proper length
      if (!Array.isArray(separators))
      {
        valid = false;
      }

      if(!valid)
      {
        $("#warning").after('<div id="warning" data-alert class="alert-box alert radius">Incorrect JSON format.<a href="#" class="close">&times;</a></div>');
      }
      else
      {
        //initialize the fictional time
        time = new fictionalTime(name, units, beginning, separators, declaration);

        //initialize the rest of the page
        startTime();
      }
    }
  });
});

/**
 * Starts all the clocks and converters
 * @function startTime
 */
function startTime()
{
  //get how long does it take for the smallest unit to elapse to set timeouts
  var timeout = time.units[time.units.length-1];

  //put the name into the text
  $("[class^=timeName]").html(time.name);

  //clocks
  localTime();
  currentFictional(timeout);
  countdownToBegin(timeout);

  //converters
  $("#eyToTimeY").submit(function(event){
    event.preventDefault();
    $("#eyToTimeYResult").text( earthYearsToTimeYears($("#eyToTimeYInput").val()) );
  });

  //SUT year to Earth year
  $("#timeToEY").submit(function(event){
    event.preventDefault();
    $("#timeToEYResult").text( timeYearsToEarthYears($("#timeToEYInput").val()) );
  });

  //setup date picker for Earth year
  $("#eDToTimeInput").fdatepicker({format:"yyyy-mm-dd"});
  //Earth date to SUT
  $("#eDToTime").submit(function(event){
    event.preventDefault();
    $("#eDToTimeResult").text( earthDateToTime($("#eDToTimeInput").val()) );
  });

}

function localTime()
{
  $("#localTime").text(moment().format("MMMM DD, YYYY HH:mm:ss"));
  setTimeout(localTime, 1000);
}

function currentFictional(timeout)
{
  $("#currentFictional").text(time.currentTime);
  setTimeout(currentFictional, timeout);
}

function countdownToBegin(timeout)
{
  $("#countdownToBegin").text(time.toTime(time.beginning - Date.now()));
  setTimeout(countdownToBegin, timeout);
}

function earthYearsToTimeYears(input)
{
  //break it down to milliseconds
  var milliseconds = input * 31557600000;

  //reconstruct into SUT and return
  return time.toFictionalUnit(milliseconds, 0);
}

function timeYearsToEarthYears(input)
{
  //first get down to milliseconds
  var milliseconds = time.unitToMilliseconds(input, 0);

  return milliseconds / 31557600000;
}

function earthDateToTime(input)
{
  return time.toDate(moment(input).valueOf());
}


js/create.js
$(function(){
  //setup date picker for beginning date
  $("#beginning").fdatepicker({format:"yyyy-mm-dd"});

  /**
   * Add unit
   */
  $("#addUnit").click(function(event){
    event.preventDefault();

    //find the latest unit
    var addPoint = $("[class^=unit]").last();
    //console.log(addPoint);

    //determine the latest unit
    var latest = addPoint.attr("id").split("-");
    latest = latest[1];
    addPoint = addPoint.parent();
    //console.log(latest);
    var unitNew = parseInt(latest) + 1;

    //for now limit to 9 units in order to prevent too much craziness
    if(unitNew < 10)
    {
      //add separator
      addPoint.after('<label>Separator '+ unitNew +'<input id="separator-'+ latest +'" class="separator" type="text" max="3" required></label>');

      //add unit
      //console.log(unitNew);
      var addPoint2 = $("[class^=separator]").last().parent();
      addPoint2.after('<label>Unit '+ unitNew +' (how many of unit '+ latest +' are in this unit)<input id="unit-'+ unitNew +'" class="unit" type="number" step="1" min="1" required></label>');
    }
    else
    {
      addPoint.after('<div data-alert class="alert-box warning">That is enough for now. If you really have such complex time, contact me (@storyteller_cz) to get around this. I\'m really curious about your universe now.</div>');
    }
  });

  /**
   * Creat the time
   */
  $("#createTime").submit(function(event){
    event.preventDefault();

    //start animation on the button
    $("#cog").attr("class", "fa fa-cog fa-spin");

    //find the total number of units
    var units = $("[class^=unit]");
    //console.log(units);
    var separators = $("[class^=separator]");
    //console.log(separators);
    var totalUnits = units.last().attr("id").split("-");
    totalUnits = parseInt(totalUnits[1]);
    //console.log(totalUnits);

    var unitsArray = new Array(totalUnits+1)
    var separatorsArray = new Array(totalUnits+2);
    //get all the units and separators into arrays
    var i = totalUnits;
    units.each(function(index, item){
      //console.log(item);
      unitsArray[i] = parseInt($(item).val());
      i--;
    });

    i = totalUnits;
    separators.each(function(index, item){
      //console.log(item);
      if(i === totalUnits)
      {
        separatorsArray[0] = $(item).val();
      }
      else
      {
        separatorsArray[i+1] = $(item).val();
      }
      i--;
    });
    //console.log(unitsArray);

    //calculate the units to milliseconds
    for (var i = 0; i < unitsArray.length; i++) {
      //add another loop to multiply by all the units bellow
      for (var k = i+1; k < unitsArray.length; k++) {
        unitsArray[i] = unitsArray[i] * unitsArray[k];
      }
    }
    //console.log(unitsArray);

    //get the remaining values
    var beginning = $("#beginning").val();
    beginning = new Date(beginning).valueOf();
    //console.log(beginning);

    var name = $("#timeName").val();
    //console.log(name);

    //merge separtorsArray with the time declaration correctly (before, after)
    var declarator = $('input[name=declaration]:checked', '#createTime').val()
    //console.log(declarator);
    if(declarator === "after")
    {
      separatorsArray[totalUnits+1] = separatorsArray[0];
      separatorsArray[0] = "";
    }
    else
    {
      separatorsArray[totalUnits+1] = "";
    }
    //console.log(separatorsArray);

    //create the JSON
    //console.log('{"name": '+JSON.stringify(name)+',"beginning": '+beginning+',"units": '+JSON.stringify(unitsArray)+',"separators": '+JSON.stringify(separatorsArray)+',"declaration": '+JSON.stringify(declarator)+'}');
    var json = JSON.parse('{"name": '+JSON.stringify(name)+',"beginning": '+beginning+',"units": '+JSON.stringify(unitsArray)+',"separators": '+JSON.stringify(separatorsArray)+',"declaration": '+JSON.stringify(declarator)+'}');
    console.log(json);

    //give JSON to the user
    $("#result").html(JSON.stringify(json));

    //stop cog spinning
    $("#cog").attr("class", "fa fa-cog");

    //move the view to the result
    $(window).scrollTop($("#result").offset().top);
  });
});


js/fictionaltime.js
/**
 * Fictional time library
 * This library works with provided fictional time and returns the fictional time
 * from provided milliseconds.
 * @author Jan Dvorak IV.
 * @copyright All rights reserved
 * @version 2.1
 * @since 2.0
 * @todo right now this class can only handle simetric times, will need to improve to handle more diverse times (test on Mayan calendar, then Augustine calendar - the main issue is the length of months)
 */

/**
 * The main object
 * @param {string} name Name of the time
 * @param {array} units Array of units lengths (int)
 * @param {array} unitsSpace how many spaces does the unit take (int)
 * @param {int} beginning The Earth Day the time beginnins in milliseconds - @todo make optional - ie. can be null
 * @param {array} separators Separations between the units
 * @param {string} declaration Where should be the time declaration displayed - before, after, both, false
 */
function fictionalTime(name, units, beginning, separators, declaration){
  //all inserted variables are public
  /** @todo revisit */
  this.name = name;
  this.units = units;
  this.beginning = beginning;
  this.separators = separators;
  this.declaration = declaration;

  //calculate unit spaces for adding appropriate number of zeroes
  var unitLength = [];
  for (var i = 0; i < units.length; i++) {
    if(i === 0)
    {
      unitLength[i] = 0;
    }
    else
    {
      var count = (units[i-1] / units[i]).toString();
      if(count[0] == 1)
      {
        //account for symetric times with  10, 100, etc.
        unitLength[i] = count.length - 1;
      }
      else
      {
        unitLength[i] = count.length;
      }
    }
  }

  /**
   * Calculates the time from milliseconds
   * @method toTime
   * @since 1.0
   * @param {int} milliseconds milliseconds that should be translated into the fictional time.
   * @return {string} Formated time display
   * @todo Add param to determine if to display declaration or not or only at one place (ie. before, after)
   */
  this.toTime = function(milliseconds)
  {
    var output = calculate(milliseconds, false);

    //return the string to display the time
    return format(output);
  }

  /**
   * Calculates the time from milliseconds and takes into account if the time
   * is before the beginning of the fictional time or not.
   * @method toDate
   * @since 2.0
   * @param {int} milliseconds milliseconds that should be translated into the fictional time.
   * @return {string} Formated time display
   * @todo Add param to determine if to display declaration or not or only at one place (ie. before, after)
   */
  this.toDate = function(milliseconds)
  {
    var output = calculate(milliseconds, true);

    //return the string to display the time
    return format(output);
  }

  /**
   * Current time
   * @method current
   * @since 2.0
   * @return {string} Formated current time in the fictional time
   */
  this.currentTime = function()
  {
    //first get current time in milliseconds
    var now = Date.now();
    //then get the offset
    var offset = new Date().getTimezoneOffset() * 60000;
    //calculate current time
    var output = calculate(new Date(now + offset), true);

    return format(output);
  }

  /**
   * Calculate specific unit from Earth Time to fictional time
   * @method toFictionalUnit
   * @since 2.0
   * @param {int} milliseconds
   * @param {int} toUnit To which unit should be the conversion done
   * @return {int} The calculated number
   */
 this.toFictionalUnit = function(milliseconds, toUnit)
 {
   return milliseconds / units[toUnit];
 }

 /**
  * Calculate specific unit from fictional time to milliseconds
  * @method unitToMilliseconds
  * @since 2.0
  * @param {int} count Number of units
  * @param {int} unit From what unit are we doing the conversion
  * @return {int} The calculated number in milliseconds
  */
  this.unitToMilliseconds = function(count, unit)
  {
    return count * units[unit];
  }

  /**
   * Calculates the fictional time
   * @method calculate
   * @since 2.0
   * @param {int} milliseconds The time to be calculated
   * @param {boolean} preTime Should the time be evaluated if it is before the establishment of the time?
   * @return {array} Array with the results to be used for the format function
   */
  function calculate(milliseconds, preTime)
  {
    var output = [];

    //evaluate preTime
    var beforeTime = false;

    //fix crazy values
    if(milliseconds < 0)
    {
      milliseconds = Math.abs(milliseconds);
    }

    //console.log("At start: " + milliseconds);
    if(preTime)
    {
      if(beginning > milliseconds)
      {
        beforeTime = true;
        //for time before establishment make it a countdown
        milliseconds = beginning - milliseconds;
      }
      else
      {
        //for a date after the establishment time remove the time to the establishment time for correct calculations
        milliseconds = milliseconds - beginning;
      }
    }

    for (var i = 0; i < units.length; i++)
    {
      //first calculate how many units are there
      var count = Math.floor( Math.abs( milliseconds / units[i]) );

      //calculate what is the maximum of the given unit
      var max = units[i-1] / units[i];

      /**
       * Assuming that the biggest unit is equivalent to years it should be counting down.
       * Other units should be going up as normal clocks.
       */
      if(i === 0 || !beforeTime)
      {
        //calculate how much of the given unit is there in the time
        output[i] = count;

        //account for getting the max unit
        if(count === max)
        {
          output[i] = 0;
          output[i-1] = parseInt(output[i-1]) + 1;
        }
      }
      else
      {
        //now get the correct number that is going to be increasing
        output[i] = max-count;
        //account for getting the max number displayed
        if(count === max)
        {
          output[i] = 0;
          output[i-1] = parseInt(output[i-1]) - 1;
        }
      }
      //reduce the time by the unit calculated
      milliseconds = milliseconds - (count * units[i]);
    }

    //have to have another loop due to lines 187 and 198
    for (var i = 0; i < units.length; i++)
    {
      //add the appropriate number of zeroes
      if(i === 0 && beforeTime)
      {
        output[i] = defaultZeros(output[i], unitLength[i], true);
      }
      else
      {
        output[i] = defaultZeros(output[i], unitLength[i], false);
      }
      //console.log("Unit " + i + " reminder after: " + milliseconds);
    }

    return output;
  }

  /**
   * Account for default zeros in the given fields
   * @method defaultZeros
   * @since 1.0
   * @param {int} number Number that should get the default zeroes before it in order to hold format
   * @param {int} unitLength How long is the given unit 1 through 3
   * @param {boolean} minus Boolean to determine if to add minus before the number or not
   * @return {string} the number with added default zeroes
   * @todo Improve that unitLength can handle more then size 3, use length of the current number to determine how many zeroes need to be added before
   */
  function defaultZeros(number, uLength, minus){

    number = number.toString();

    //first determine how many zeroes need to be added
    var add = uLength - number.length;

    //add the zeroes before the given number
    for (var i = 0; i < add; i++) {

      number = "0" + number;
    }
    //account for minus
    if(minus)
    {
      number = "-" + number;
    }

    return number;
  }

  /**
   * Function to return the time in the correct format
   * @method format
   * @since 1.1
   * @param {array} array with the values for each fields
   * @return {string}
   */
   function format(valuesArray)
   {
     //return the string to display the time
     var outputString;
     for (var i = 0; i < valuesArray.length; i++) {
       var k = i+1;

       //unit declaration before time
       if(i === 0)
       {
         if(declaration === "before" || declaration === "both")
         {
           outputString = separators[i];
         }
       }

       outputString += valuesArray[i] + separators[k];

       //unit declaration after time
       if(i === valuesArray.length)
       {
         if(declaration === "after" || declaration === "both")
         {
           outputString += separators[k+1];
         }
       }

     }
     return outputString;
   }

//end of the class
}


js/main.js
/**
 * Struggle for Balance time conventor
 * @author Jan Dvorak IV.
 */

/*
 * Settings
 * Standard unit is milisecond from which all calculations are going to be made
 */
var unit1 = 1000; //how many miliseconds does the unit have
var unit2 = unit1 * 100;
var unit3 = unit2 * 100;
var unit4 = unit3 * 10;
var unit5 = unit4 * 500;

var unitArray = [unit5, unit4, unit3, unit2, unit1]; //must be in order from the largest to smallest
var unitSeparator = ["SUT ", ".", " ", ":", ":", ""]; //formating, determines what symbols will be between different
var unitDeclaration = "before"; //options: before, after, both, false - additional option for unitSeparator to show time declaration like SUT xxxx

var timeBegins = moment("2400/01/01 00:00 GMT"); //the moment when time begins in relation to Earth time

//create the object
var sut = new fictionalTime("Standard Universal Time", unitArray, timeBegins.valueOf(), unitSeparator, unitDeclaration);

//old variables
var sutMinuteSec = 100;
var sutHourSec = sutMinuteSec * 100;
var sutDaySec = sutHourSec * 10;
var sutYearSec = sutDaySec * 500;

var earthYearReformed = 31622400000;

$(function(){
  //time to sut start
  $("#toSut").text(timeBegins.fromNow());

  //today at GMT
  todayGMT();
  //today at local time
  todayLocal();
  //countdown to SUT in SUT
  countdownToBegin();
  //passed SUT
  passedSUT();
  //curent date and time
  curentSUT();

  //forms
  //Earth year to SUT year
  $("#eyToSUTY").submit(function(event){
    event.preventDefault();
    $("#eyToSUTYResult").text( earthYearsToSUT($("#eyToSUTYInput").val()) );
  });

  //SUT year to Earth year
  $("#sutToEY").submit(function(event){
    event.preventDefault();
    $("#sutToEYResult").text( SUTyearsToEarth($("#sutToEYInput").val()) );
  });

  //setup date picker for Earth year
  $("#eDToSUTInput").fdatepicker({format:"yyyy-mm-dd"});
  //Earth date to SUT
  $("#eDToSUT").submit(function(event){
    event.preventDefault();
    $("#eDToSUTResult").text( earthDateToSUT($("#eDToSUTInput").val()) );
  });

  //SUT date to Earth date
  $("#sutToED").submit(function(event){
    event.preventDefault();
    $("#sutToEDResult").text( sutDateToEY($("#sutToEDInput").val()) );
  });
});

/**
 * Shows current GMT time
 */
function todayGMT(){
  $("#todayGMT").text(moment().utc().format("MMMM DD, YYYY HH:mm:ss"));
  setTimeout(todayGMT, 1000);
}

/**
 * Shows current local time
 */
function todayLocal(){
  $("#todayLocal").text(moment().format("MMMM DD, YYYY HH:mm:ss"));
  setTimeout(todayLocal, 1000);
}

/**
 * Calculates current SUT time for the local time
 */
function countdownToBegin(){
  $("#countdownToBegin").text(sut.toTime(sut.beginning - Date.now()));

  setTimeout(countdownToBegin, 1000);
}

/**
 * A clock that counts how much time has passed since user opened the webpage
 */
var timeOnPage = 0;
function passedSUT(){
  timeOnPage += 1000;
  $("#passedSUT").text(sut.toTime(timeOnPage));
  setTimeout(passedSUT, 1000);
}

/**
 * Current SUT time
 */
function curentSUT(){
  $("#curentSUT").text(sut.currentTime);
  setTimeout(curentSUT, 1000);
}

/**
 * Takes in SUT years and figures out Earth years based on years after calendar reform of 2400
 */
function SUTyearsToEarth(sutYears){
  //first get down to milliseconds
  var milliseconds = sut.unitToMilliseconds(sutYears, 0);

  return milliseconds / earthYearReformed;
}

/**
 * Takes Earth years and converts them to SUT years
 */
function earthYearsToSUT(earthYear)
{
  //break it down to milliseconds
  var milliseconds = earthYear * earthYearReformed;

  //reconstruct into SUT and return
  return sut.toFictionalUnit(milliseconds, 0);
}

/**
 * Gets an Earth date and converts it to SUT date
 */
function earthDateToSUT(earthDate){
  return sut.toDate(moment.utc(earthDate).valueOf());
}


/**
 * Get SUT date and convert it to Earth date
 */
function sutDateToEY(sutDate){

  //validate we have a number
  //@todo improve validation
  if(isNaN(sutDate))
  {
    return "Please enter a valid SUT date in this format: YYY.DDD";
  }

  //validate that the input is correct
  var input = sutDate.split(".");

  if(input.length !== 2)
  {
    //there should be two entries, if there aren't return error
    return "Please enter a valid SUT date in this format: YYY.DDD";
  }

  var year = input[0];
  var day = input[1];

  //validate input
  if(day > 500)
  {
    return "SUT Year has only 500 days. Which means you should enter days between 000 to 499.";
  }

  //get the seconds for the years
  var milliseconds = sut.unitToMilliseconds(year, 0);
  //get the seconds for the days
  milliseconds += sut.unitToMilliseconds(day, 1);

  //add year 2400
  milliseconds = timeBegins.valueOf() + milliseconds;
  return moment.utc(milliseconds).format('YYYY-MM-DD');
}
